code/_onclick/hud/screen_objects.dm

```diff
@@ -89,6 +89,94 @@
+	var/has_craft_items = FALSE
+
+/atom/movable/screen/craft/Initialize(mapload)
+	. = ..()
+	register_context()
+
+/atom/movable/screen/craft/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+	. = ..()
+	context[SCREENTIP_CONTEXT_RMB] = "Display Possible Recipes"
+	return CONTEXTUAL_SCREENTIP_SET
+
+/atom/movable/screen/craft/Click(location, control, params)
+	var/list/modifiers = params2list(params)
+	if(LAZYACCESS(modifiers, RIGHT_CLICK))
+		for(var/hand_number in 0 to length(hud.mymob.held_items))
+			var/item_in_hand = hud.mymob.get_item_for_held_index(hand_number)
+			display_possible_recipes(item_in_hand)
+		if(length(hud.mymob.held_items) == length(hud.mymob.get_empty_held_indexes()))
+			to_chat(hud.mymob, span_warning("You aren't holding anything in your hands!"))
+	else
+		..()
+
+/atom/movable/screen/craft/proc/display_possible_recipes(obj/item/item)
+	if(!item)
+		return
+	var/list/possible_recipes = list()
+	var/recipe_line = ""
+	var/all_recipes = ""
+	for(var/datum/crafting_recipe/recipe in GLOB.crafting_recipes)
+		if((item.type in recipe.reqs) || ((item.parent_type in recipe.reqs) && !(item.type in recipe.blacklist)) || item.type == recipe.result)
+			possible_recipes += recipe
+
+	if(!length(possible_recipes))
+		to_chat(hud.mymob, span_warning("No recipes found for [item.name]!"))
+		return
+
+	for(var/datum/crafting_recipe/recipe in possible_recipes)
+		var/list/reqs_text = list()
+		var/list/tools_other_text = list()
+		var/list/tools_text = list()
+		for(var/obj/item/recipe_item as anything in recipe.reqs)
+			var/recipe_reagent = FALSE
+			if(ispath(recipe_item, /datum/reagent))
+				recipe_reagent = TRUE
+			var/amount = recipe.reqs[recipe_item]
+			reqs_text += "[span_notice("[initial(recipe_item.name)]")] ([span_red("[amount][recipe_reagent ? "u" : null]")])"
+		if(length(recipe.tool_paths))
+			for(var/obj/item/tool_other as anything in recipe.tool_paths)
+				tools_other_text += initial(tool_other.name)
+		if(length(recipe.tool_behaviors))
+			for(var/tool in recipe.tool_behaviors)
+				tools_text += tool
+		recipe_line = "<b>[recipe.name]</b>: [jointext(reqs_text, ", ")][(length(reqs_text) && length(tools_other_text)) ? ", " : null][span_green("[jointext(tools_other_text, ", ")][((length(tools_other_text) || length(reqs_text)) && length(tools_text))  ? ", " : null][jointext(tools_text, ", ")]")]"
+		all_recipes += "[recipe_line]\n"
+
+	to_chat(hud.mymob, examine_block("[span_ooc("Recipes for [item.name] [icon2html(item, hud.mymob)]")]\n[all_recipes]"))
+
+/atom/movable/screen/craft/update_overlays()
+	. = ..()
+	if(has_craft_items)
+		. += image('icons/hud/screen_gen.dmi', icon_state = "can_craft")
+
+/atom/movable/screen/craft/proc/hands_craft_check(mob/living/carbon/human/user, obj/item/item, slot)
+	SIGNAL_HANDLER
+	if(slot != ITEM_SLOT_HANDS)
+		for(var/obj/item/hand_item in user.held_items)
+			if(hand_item.crafting_item_in_hands && hand_item != item)
```

Comment:
so you're using `.crafting_item_in_hands` to cache that typecache lookup. smart

Reply:
rather then storing that information on the item though, why don't we store a list of weakrefs on the crafting button?
That way everything related to us is stored on us, instead of somewhere else